src/nvim/terminal.c

```diff
@@ -1184,23 +1184,31 @@ static void on_scrollback_option_changed(Terminal *term, buf_T *buf)
 static void refresh_scrollback(Terminal *term, buf_T *buf)
 {
   int width, height;
+  int appended_lines_count;
   vterm_get_size(term->vt, &height, &width);
 
-  while (term->sb_pending > 0) {
+  if (((int)buf->b_ml.ml_line_count - height) >= (int)term->sb_size) {
+    // scrollback buffer is full - delete 10% from the top
+    int sb_delete = (int)term->sb_size / 10;
+    for(int i = 1; i <= sb_delete; ++i, --term->sb_current) {
+        ml_delete(i, false);
+        xfree(term->sb_buffer[term->sb_current]);
```

Comment:
I'm not sure that `xfree` is wanted here. Note that `on_scrollback_option_changed` does that because when the actual 'scrollback' option changes we want libvterm's internal scrollback to change. But not here.

Here (in `refresh_scrollback`) we're only updating the vim buffer.

Reply:
> I'm not sure that xfree is wanted here.

I see... but the memory is still being allocated to libvterm's internal scrollback. Doesn't it bother us? Or `on_scrollback_option_changed` should take care of that?

Reply:
> but the memory is still being allocated to libvterm's internal scrollback

yes but that allocation isn't done by `refresh_scrollback` , it's done elsewhere, and also freed elsewhere, just as before this PR.

Reply:
> yes but that allocation isn't done by refresh_scrollback , it's done elsewhere, and also freed elsewhere (terminal_destroy), just as before this PR.

Thanks for explaining